File: aoc-2023/src/main/java/com/capital7software/aoc/aoc2023/days/Day09.java
package com.capital7software.aoc.aoc2023.days;

import com.capital7software.aoc.lib.AdventOfCodeSolution;
import com.capital7software.aoc.lib.math.HistorySequence;
import org.jetbrains.annotations.NotNull;

import java.time.Instant;
import java.util.List;
import java.util.logging.Logger;

/**
 * --- Day 9: Mirage Maintenance ---<br>
 * You ride the camel through the sandstorm and stop where the ghost's maps told you to stop.
 * The sandstorm subsequently subsides, somehow seeing you standing at an oasis!
 * <p><br>
 * The camel goes to get some water, and you stretch your neck. As you look up, you discover
 * what must be yet another giant floating island, this one made of metal! That must be where
 * the parts to fix the sand machines come from.
 * <p><br>
 * There's even a hang glider partially buried in the sand here; once the sun rises and heats
 * up the sand, you might be able to use the glider and the hot air to get all the way up to the metal island!
 * <p><br>
 * While you wait for the sun to rise, you admire the oasis hidden here in the middle of
 * Desert Island. It must have a delicate ecosystem; you might as well take some ecological
 * readings while you wait. Maybe you can report any environmental instabilities you find to
 * someone so the oasis can be around for the next sandstorm-worn traveler.
 * <p><br>
 * You pull out your handy Oasis And Sand Instability Sensor and analyze your surroundings.
 * The OASIS produces a report of many values and how they are changing over time (your puzzle input).
 * Each line in the report contains the history of a single value. For example:
 * <p><br>
 * <code>
 * 0 3 6 9 12 15<br>
 * 1 3 6 10 15 21<br>
 * 10 13 16 21 30 45<br>
 * </code>
 * <p><br>
 * To best protect the oasis, your environmental report should include a prediction of the next value
 * in each history. To do this, start by making a new sequence from the difference at each step of
 * your history. If that sequence is not all zeroes, repeat this process, using the sequence you
 * just generated as the input sequence. Once all of the values in your latest sequence are zeroes,
 * you can extrapolate what the next value of the original history should be.
 * <p><br>
 * In the above dataset, the first history is 0 3 6 9 12 15. Because the values increase by 3 each step,
 * the first sequence of differences that you generate will be 3 3 3 3 3. Note that this sequence has
 * one fewer value than the input sequence because at each step it considers two numbers from the input.
 * Since these values aren't all zero, repeat the process: the values differ by 0 at each step,
 * so the next sequence is 0 0 0 0. This means you have enough information to extrapolate the history!
 * Visually, these sequences can be arranged like this:
 * <p><br>
 * <code>
 * 0   3   6   9  12  15<br>
 * 3   3   3   3   3<br>
 * 0   0   0   0<br>
 * </code>
 * <p><br>
 * To extrapolate, start by adding a new zero to the end of your list of zeroes; because the zeroes
 * represent differences between the two values above them, this also means there is now a placeholder
 * in every sequence above it:
 * <p><br>
 * <code>
 * 0   3   6   9  12  15   B<br>
 * 3   3   3   3   3   A<br>
 * 0   0   0   0   0<br>
 * </code>
 * <p><br>
 * You can then start filling in placeholders from the bottom up. A needs to be the result of
 * increasing 3 (the value to its left) by 0 (the value below it); this means A must be 3:
 * <p><br>
 * <code>
 * 0   3   6   9  12  15   B<br>
 * 3   3   3   3   3   3<br>
 * 0   0   0   0   0<br>
 * </code>
 * <p><br>
 * Finally, you can fill in B, which needs to be the result of increasing 15 (the value to its left)
 * by 3 (the value below it), or 18:
 * <p><br>
 * <code>
 * 0   3   6   9  12  15  18<br>
 * 3   3   3   3   3   3<br>
 * 0   0   0   0   0<br>
 * </code>
 * <p><br>
 * So, the next value of the first history is 18.
 * <p><br>
 * Finding all-zero differences for the second history requires an additional sequence:
 * <p><br>
 * <code>
 * 1   3   6  10  15  21<br>
 * 2   3   4   5   6<br>
 * 1   1   1   1<br>
 * 0   0   0<br>
 * </code>
 * <p><br>
 * Then, following the same process as before, work out the next value in each sequence from the bottom up:
 * <p><br>
 * <code>
 * 1   3   6  10  15  21  28<br>
 * 2   3   4   5   6   7<br>
 * 1   1   1   1   1<br>
 * 0   0   0   0<br>
 * </code>
 * <p><br>
 * So, the next value of the second history is 28.
 * <p><br>
 * The third history requires even more sequences, but its next value can be found the same way:
 * <p><br>
 * <code>
 * 10  13  16  21  30  45  68<br>
 * 3   3   5   9  15  23<br>
 * 0   2   4   6   8<br>
 * 2   2   2   2<br>
 * 0   0   0<br>
 * </code>
 * <p><br>
 * So, the next value of the third history is 68.
 * <p><br>
 * If you find the next value for each history in this example and add them together, you get 114.
 * <p><br>
 * Analyze your OASIS report and extrapolate the next value for each history. What is the sum of
 * these extrapolated values?
 * <p><br>
 * Your puzzle answer was 1901217887.
 * <p><br>
 * --- Part Two ---<br>
 * Of course, it would be nice to have even more history included in your report. Surely it's safe
 * to just extrapolate backwards as well, right?
 * <p><br>
 * For each history, repeat the process of finding differences until the sequence of differences
 * is entirely zero. Then, rather than adding a zero to the end and filling in the next values of
 * each previous sequence, you should instead add a zero to the beginning of your sequence of zeroes,
 * then fill in new first values for each previous sequence.
 * <p><br>
 * In particular, here is what the third example history looks like when extrapolating back in time:
 * <p><br>
 * <code>
 * 5  10  13  16  21  30  45<br>
 * 5   3   3   5   9  15<br>
 * -2   0   2   4   6<br>
 * 2   2   2   2<br>
 * 0   0   0<br>
 * </code>
 * <p><br>
 * Adding the new values on the left side of each sequence from bottom to top eventually reveals
 * the new left-most history value: 5.
 * <p><br>
 * Doing this for the remaining example data above results in previous values of -3 for the first
 * history and 0 for the second history. Adding all three new values together produces 2.
 * <p><br>
 * Analyze your OASIS report again, this time extrapolating the previous value for each history.
 * What is the sum of these extrapolated values?
 * <p><br>
 * Your puzzle answer was 905.
 */
public class Day09 implements AdventOfCodeSolution {
    private static final Logger LOGGER = Logger.getLogger(Day09.class.getName());

    /**
     * Instantiates this Solution instance.
     */
    public Day09() {

    }

    @Override
    public String getDefaultInputFilename() {
        return "inputs/input_day_09-01.txt";
    }

    @Override
    public void runPart1(List<String> input) {
        var start = Instant.now();
        var answer = sumOfNextPredictedHistories(input);
        var end = Instant.now();

        LOGGER.info(String.format("The sum of the next predicted values is: %d%n", answer));
        logTimings(LOGGER, start, end);
    }

    @Override
    public void runPart2(List<String> input) {
        var start = Instant.now();
        var answer = sumOfPreviousPredictedHistories(input);
        var end = Instant.now();

        LOGGER.info(String.format("The sum of the previous predicted values is: %d%n", answer));
        logTimings(LOGGER, start, end);
    }

    /**
     * Returns the sum of the next predicted value for all HistorySequences in the specified input.
     *
     * @param input The List of Strings of histories to parse and load.
     * @return The sum of the next predicted value for all parsed and loaded histories.
     */
    public long sumOfNextPredictedHistories(List<String> input) {
        var histories = loadHistories(input);
        return histories.stream().mapToLong(HistorySequence::predictNextValue).sum();
    }

    /**
     * Returns the sum of the previous predicted value for all HistorySequences in the specified input.
     *
     * @param input The List of Strings of histories to parse and load.
     * @return The sum of the previous predicted value for all parsed and loaded histories.
     */
    public long sumOfPreviousPredictedHistories(List<String> input) {
        var histories = loadHistories(input);
        return histories.stream().mapToLong(HistorySequence::predictPreviousValue).sum();
    }

    private @NotNull List<HistorySequence> loadHistories(List<String> input) {
        return input.stream().map(HistorySequence::new).toList();
    }
}

File: aoc-2017/src/main/kotlin/com/capital7software/aoc/aoc2017aoc/days/Day09.kt
package com.capital7software.aoc.aoc2017aoc.days

import com.capital7software.aoc.lib.AdventOfCodeSolution
import com.capital7software.aoc.lib.string.StreamProcessing
import java.time.Instant
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * **--- Day 9: Stream Processing ---**
 *
 * A large stream blocks your path. According to the locals, it's not safe to cross the
 * stream at the moment because it's full of **garbage**. You look down at the stream; rather
 * than water, you discover that it's a **stream of characters.**
 *
 * You sit for a while and record part of the stream (your puzzle input). The characters
 * represent **groups** - sequences that begin with { and end with }. Within a group, there are
 * zero or more other things, separated by commas: either another **group** or **garbage**. Since
 * groups can contain other groups, a } only closes the **most-recently-opened unclosed
 * group** - that is, they are nestable. Your puzzle input represents a single, large group
 * which itself contains many smaller ones.
 *
 * Sometimes, instead of a group, you will find **garbage**. Garbage begins with < and
 * ends with >. Between those angle brackets, almost any character can appear,
 * including { and }. **Within** garbage, < has no special meaning.
 *
 * In a futile attempt to clean up the garbage, some program has **canceled** some of the
 * characters within it using !: inside garbage, **any** character that comes
 * after ! should be **ignored**, including <, >, and even another !.
 *
 * You don't see any characters that deviate from these rules. Outside garbage,
 * you only find well-formed groups, and garbage always terminates according to the rules above.
 *
 * Here are some self-contained pieces of garbage:
 *
 * - ```<>```, empty garbage.
 * - ```<random characters>```, garbage containing random characters.
 * - ```<<<<>```, because the extra < are ignored.
 * - ```<{!>}>```, because the first > is canceled.
 * - ```<!!>```, because the second ! is canceled, allowing the > to terminate the garbage.
 * - ```<!!!>>```, because the second ! and the first > are canceled.
 * - ```<{o"i!a,<{i<a>```, which ends at the first >.
 *
 * Here are some examples of whole streams and the number of groups they contain:
 *
 * - ```{}```, 1 group.
 * - ```{{{}}}```, 3 groups.
 * - ```{{},{}}```, also 3 groups.
 * - ```{{{},{},{{}}}}```, 6 groups.
 * - ```{<{},{},{{}}>}```, 1 group (which itself contains garbage).
 * - ```{<a>,<a>,<a>,<a>}```, 1 group.
 * - ```{{<a>},{<a>},{<a>},{<a>}}```, 5 groups.
 * - ```{{<!>},{<!>},{<!>},{<a>}}```, 2 groups (since all but the last > are canceled).
 *
 * Your goal is to find the total score for all groups in your input. Each group is
 * assigned a score which is one more than the score of the group that immediately
 * contains it. (The outermost group gets a score of 1.)
 *
 * - ```{}```, score of 1.
 * - ```{{{}}}```, score of 1 + 2 + 3 = 6.
 * - ```{{},{}}```, score of 1 + 2 + 2 = 5.
 * - ```{{{},{},{{}}}}```, score of 1 + 2 + 3 + 3 + 3 + 4 = 16.
 * - ```{<a>,<a>,<a>,<a>}```, score of 1.
 * - ```{{<ab>},{<ab>},{<ab>},{<ab>}}```, score of 1 + 2 + 2 + 2 + 2 = 9.
 * - ```{{<!!>},{<!!>},{<!!>},{<!!>}}```, score of 1 + 2 + 2 + 2 + 2 = 9.
 * - ```{{<a!>},{<a!>},{<a!>},{<ab>}}```, score of 1 + 2 = 3.
 *
 * **What is the total score** for all groups in your input?
 *
 * Your puzzle answer was **7640**.
 *
 * **--- Part Two ---**
 *
 * Now, you're ready to remove the garbage.
 *
 * To prove you've removed it, you need to count all the characters within the garbage.
 * The leading and trailing < and > don't count, nor do any canceled
 * characters or the ! doing the canceling.
 *
 * - <>, 0 characters.
 * - <random characters>, 17 characters.
 * - <<<<>, 3 characters.
 * - <{!>}>, 2 characters.
 * - <!!>, 0 characters.
 * - <!!!>>, 0 characters.
 * - <{o"i!a,<{i<a>, 10 characters.
 *
 * **How many non-canceled characters are within the garbage** in your puzzle input?
 *
 * Your puzzle answer was **4368**.
 */
class Day09 : AdventOfCodeSolution {
  companion object {
    private val log: Logger = LoggerFactory.getLogger(Day09::class.java)
  }

  override fun getDefaultInputFilename(): String = "inputs/input_day_09-01.txt"

  override fun runPart1(input: MutableList<String>) {
    val start = Instant.now()
    val answer = scoreOfStream(input.first())
    val end = Instant.now()

    log.info("$answer is the score of all the groups in the stream!")
    logTimings(log, start, end)
  }

  override fun runPart2(input: MutableList<String>) {
    val start = Instant.now()
    val answer = garbageRemoved(input.first())
    val end = Instant.now()

    log.info("$answer is the number of characters removed while in garbage!")
    logTimings(log, start, end)
  }

  /**
   * Returns the sum of all the group scores in the stream.
   *
   * @param input The [String] to stream.
   * @return The sum of all the group scores in the stream.
   */
  fun scoreOfStream(input: String): Int {
    val instance = StreamProcessing(input)
    val score = instance.score
    return score.second
  }

  /**
   * Returns the number of characters removed while in garbage.
   *
   * @param input The [String] to stream.
   * @return The number of characters removed while in garbage.
   */
  fun garbageRemoved(input: String): Int {
    val instance = StreamProcessing(input)
    val score = instance.score
    return score.first
  }
}
